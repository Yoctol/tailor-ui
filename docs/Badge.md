---
id: badge
title: Badge
---

Small numerical value or status descriptor for UI elements.

## When To Use

Badge normally appears in proximity to notifications or user avatars with eye-catching appeal, typically displaying unread messages count.

## Examples

```js
import { Badge } from 'tailor-ui';
```

### Basic

```jsx live
<Badge count={2}>
  <Button variant="normal" icon={MdNotifications} />
</Badge>
```

### Standalone

```jsx live
<Flex alignItems="center">
  <Heading.h4 mr="2px">Notifications</Heading.h4>
  <Badge count={2} />
</Flex>
```

### Customized color

```jsx live
<>
  <Badge count={16} bg="success" wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={100} bg="light" color="gray500" borderColor="gray100">
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
</>
```

### Overflow count

```jsx live
<>
  <Badge count={99} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={100} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={99} overflowCount={10} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={1000} overflowCount={999} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
</>
```

## API

| Property        | Description             | Type     | Default |
|-----------------|-------------------------|----------|---------|
| `count`         | Number to show in badge | `number` |         |
| `overflowCount` | Max count to show       | `number` | 99      |
| `wrapperProps`  | Props of badge wrapper  | BoxProps |         |
| `color`         | Count color             | string   |         |
| `bg`            | background color        | string   |         |
| `borderColor`   | border color            | string   |         |
