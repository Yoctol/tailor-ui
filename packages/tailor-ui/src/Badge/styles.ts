import styled from 'styled-components';
import { ComponentPropsWithRef, ForwardRefExoticComponent } from 'react';
import { animated } from 'react-spring';

import { Box } from '../Layout';

const StyledBadge = styled(Box)`
  display: inline-flex;
  align-items: center;
  min-width: 16px;
  height: 16px;
  padding: 0 4px;
  border: ${(p) => p.theme.borders.base};
  border-radius: ${(p) => p.theme.radii.base};
  font-size: 12px;
  line-height: 1;
`;

export const AnimatedStyledBadge: ForwardRefExoticComponent<ComponentPropsWithRef<
  typeof StyledBadge
>> = animated(StyledBadge);

export const StyledBadgeWrapper = styled(Box)`
  display: inline-flex;
  position: relative;

  ${StyledBadge} {
    position: absolute;
    top: 0;
    right: 8px;
  }
`;
