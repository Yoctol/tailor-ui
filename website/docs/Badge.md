---
id: badge
title: Badge
---

Small numerical value or status descriptor for UI elements.

## When To Use

Badge normally appears in proximity to notifications or user avatars with eye-catching appeal, typically displaying unread messages count.

## Examples

```js
import { Badge } from 'tailor-ui';
```

### Basic

```jsx live
<Badge count={2}>
  <Button variant="normal" icon={MdNotifications} />
</Badge>
```

### Standalone

```jsx live
<Flex alignItems="center">
  <Heading.H4 mr="2px">Notifications</Heading.H4>
  <Badge count={2} />
</Flex>
```

### Customized color

```jsx live
<>
  <Badge count={16} bg="success" wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={100} bg="light" color="gray500" borderColor="gray100">
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
</>
```

### Overflow count

```jsx live
<>
  <Badge count={99} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={100} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={99} overflowCount={10} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
  <Badge count={1000} overflowCount={999} wrapperProps={{ mr: '3' }}>
    <Button variant="normal" icon={MdNotifications} />
  </Badge>
</>
```

### Dynamic count

```jsx live
() => {
  const [count, setCount] = useState(0);

  return (
    <div>
      <Badge count={count}>
        <Button variant={count === 0 ? undefined: 'normal'} icon={MdNotifications} />
      </Badge>
      <br />
      <br />
      <Button
        width="36px"
        size="sm"
        onClick={() => setCount(count + 1)}
      >
        +
      </Button>
      <Button
        ml="2"
        width="36px"
        size="sm"
        onClick={() => {
          if (count > 0) {
            setCount(count - 1);
          }
        }}
      >
        -
      </Button>
    </div>
  );
}
```


## API

| Property        | Description                              | Type      | Default |
|-----------------|------------------------------------------|-----------|---------|
| `count`         | Number to show in badge                  | `number`  |         |
| `overflowCount` | Max count to show                        | `number`  | 99      |
| `showZero`      | Whether to show badge when count is zero | `boolean` | `false` |
| `wrapperProps`  | Props of badge wrapper                   | BoxProps  |         |
| `color`         | Count color                              | string    |         |
| `bg`            | background color                         | string    |         |
| `borderColor`   | border color                             | string    |         |
