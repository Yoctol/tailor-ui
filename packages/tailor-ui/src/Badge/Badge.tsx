import React, { ComponentPropsWithRef, FC } from 'react';

import { StyledBadge, StyledBadgeWrapper } from './styles';

type StyledBadgeProps = ComponentPropsWithRef<typeof StyledBadge>;

export type BadgeProps = StyledBadgeProps & {
  count?: number;
  overflowCount?: number;
  wrapperProps?: ComponentPropsWithRef<typeof StyledBadgeWrapper>;
};

const Badge: FC<BadgeProps> = ({
  count,
  overflowCount = 99,
  wrapperProps,
  children,
  color = 'light',
  bg = 'danger',
  borderColor = 'light',
  ...props
}) => {
  const displayCount =
    count && count > overflowCount ? `${overflowCount}+` : count;

  const badge = (
    <StyledBadge color={color} bg={bg} borderColor={borderColor} {...props}>
      {displayCount}
    </StyledBadge>
  );

  if (!children) {
    return badge;
  }

  return (
    <StyledBadgeWrapper {...wrapperProps}>
      {children}
      {badge}
    </StyledBadgeWrapper>
  );
};

export { Badge };
