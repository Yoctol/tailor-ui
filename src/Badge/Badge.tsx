import React, { ComponentPropsWithRef, FC } from 'react';
import { to, useTransition } from '@react-spring/web';

import { AnimatedStyledBadge, StyledBadgeWrapper } from './styles';

type StyledBadgeProps = ComponentPropsWithRef<typeof AnimatedStyledBadge>;

export type BadgeProps = StyledBadgeProps & {
  count?: number;
  showZero?: boolean;
  overflowCount?: number;
  wrapperProps?: ComponentPropsWithRef<typeof StyledBadgeWrapper>;
};

const Badge: FC<BadgeProps> = ({
  count,
  overflowCount = 99,
  showZero = false,
  wrapperProps,
  children,
  color = 'light',
  bg = 'danger',
  borderColor = 'light',
  ...otherProps
}) => {
  const displayCount =
    count && count > overflowCount ? `${overflowCount}+` : count;
  const transition = useTransition(showZero || count !== 0, {
    from: {
      opacity: 0,
      transform: 'scale(0.6)',
    },
    enter: {
      opacity: 1,
      transform: 'scale(1)',
    },
    leave: {
      opacity: 0,
      transform: 'scale(0.6)',
    },
  });

  const badge = transition(
    (style, item) =>
      item && (
        <AnimatedStyledBadge
          color={color}
          bg={bg}
          borderColor={borderColor}
          {...otherProps}
          style={{
            opacity: style.opacity,
            transformOrigin: children ? 'right' : 'center',
            transform: to(style.transform, (x) =>
              children ? `${x} translateX(50%)` : x
            ),
          }}
        >
          {displayCount}
        </AnimatedStyledBadge>
      )
  );

  if (!children) {
    return badge;
  }

  return (
    <StyledBadgeWrapper {...wrapperProps}>
      {children}
      {badge}
    </StyledBadgeWrapper>
  );
};

export { Badge };
