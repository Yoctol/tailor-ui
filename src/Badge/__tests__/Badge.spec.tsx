import React, { useState } from 'react';
import userEvent from '@testing-library/user-event';

import { render, screen, waitFor } from 'test/test-utils';

import { Badge } from '../Badge';

describe('Backdrop', () => {
  it('should render badge correctly', async () => {
    const { container } = render(
      <Badge count={2}>
        <button type="button">btn</button>
      </Badge>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render standalone badge correctly', async () => {
    const { container } = render(<Badge count={2} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render customized badge correctly', async () => {
    const { container } = render(
      <Badge count={16} bg="success">
        <button type="button">btn</button>
      </Badge>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render overflowCount badge correctly', async () => {
    const { container } = render(
      <Badge count={100}>
        <button type="button">btn</button>
      </Badge>
    );

    expect(container).toHaveTextContent('99+');
  });

  it('should render customized overflowCount badge correctly', async () => {
    const { container } = render(
      <Badge count={1000} overflowCount={999}>
        <button type="button">btn</button>
      </Badge>
    );

    expect(container).toHaveTextContent('999+');
  });

  it('should render badge with 0 count when showZero is true', async () => {
    const { container } = render(
      <Badge count={0} showZero>
        <button type="button">btn</button>
      </Badge>
    );

    expect(container).toHaveTextContent('0');
  });

  it('should not render badge with 0 by default', async () => {
    const { container } = render(
      <Badge count={0}>
        <button type="button">btn</button>
      </Badge>
    );

    expect(container).not.toHaveTextContent('0');
  });

  it('should render badge dynamically', async () => {
    const DynamicBadge = () => {
      const [count, setCount] = useState(0);

      return (
        <div>
          <Badge count={count}>
            <button type="button">btn</button>
          </Badge>
          <button type="button" onClick={() => setCount((prev) => prev + 1)}>
            +
          </button>
          <button type="button" onClick={() => setCount((prev) => prev - 1)}>
            -
          </button>
        </div>
      );
    };

    const { container } = render(<DynamicBadge />);

    const incrementBtn = screen.getByText('+');
    const decrementBtn = screen.getByText('-');

    expect(container).not.toHaveTextContent('0');

    userEvent.click(incrementBtn);
    expect(container).toHaveTextContent('1');

    userEvent.click(decrementBtn);
    await waitFor(() => expect(container).not.toHaveTextContent('0'));
  });
});
