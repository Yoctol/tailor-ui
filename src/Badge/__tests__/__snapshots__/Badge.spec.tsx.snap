// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Backdrop should render badge correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
  color: #ffffff;
  background-color: #ff5757;
  border-color: #ffffff;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 16px;
  height: 16px;
  padding: 0 4px;
  border: 1px solid;
  border-radius: 0.25rem;
  font-size: 12px;
  line-height: 1;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 .c2 {
  position: absolute;
  top: 0;
  right: 8px;
}

<div
  class="c0 c1"
>
  <button
    type="button"
  >
    btn
  </button>
  <div
    class="c2 c3 c4"
    color="light"
    style="opacity: 1; transform-origin: right; transform: scale(1) translateX(50%);"
  >
    2
  </div>
</div>
`;

exports[`Backdrop should render customized badge correctly 1`] = `
.c0 {
  box-sizing: border-box;
}

.c3 {
  box-sizing: border-box;
  color: #ffffff;
  background-color: #51c81c;
  border-color: #ffffff;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 16px;
  height: 16px;
  padding: 0 4px;
  border: 1px solid;
  border-radius: 0.25rem;
  font-size: 12px;
  line-height: 1;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  position: relative;
}

.c1 .c2 {
  position: absolute;
  top: 0;
  right: 8px;
}

<div
  class="c0 c1"
>
  <button
    type="button"
  >
    btn
  </button>
  <div
    class="c2 c3 c4"
    color="light"
    style="opacity: 1; transform-origin: right; transform: scale(1) translateX(50%);"
  >
    16
  </div>
</div>
`;

exports[`Backdrop should render standalone badge correctly 1`] = `
.c0 {
  box-sizing: border-box;
  color: #ffffff;
  background-color: #ff5757;
  border-color: #ffffff;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 16px;
  height: 16px;
  padding: 0 4px;
  border: 1px solid;
  border-radius: 0.25rem;
  font-size: 12px;
  line-height: 1;
}

<div
  class="c0 c1"
  color="light"
  style="opacity: 1; transform-origin: center; transform: scale(1);"
>
  2
</div>
`;
